sdc-plugin/clocks.cc

```diff
@@ -0,0 +1,255 @@
+		PropagateThroughBuffer(pass, pll_clock, Bufg());
+	    }
+	}
+    }
+#ifdef SDC_DEBUG
+    log("Finish clock divider clock propagation\n\n");
+#endif
+}
+
+void Clocks::PropagateThroughBuffer(Propagation* pass, Clock& clock,
+                                    Buffer buffer) {
+    auto clock_wires = clock.GetClockWires();
+    for (auto clock_wire : clock_wires) {
+#ifdef SDC_DEBUG
+	log("Clock wire %s\n", RTLIL::unescape_id(clock_wire->name).c_str());
+#endif
+	auto buf_wires = pass->FindSinkWiresForCellType(clock_wire, buffer.name,
+	                                                buffer.output);
+	int path_delay(0);
+	for (auto wire : buf_wires) {
+#ifdef SDC_DEBUG
+	    log("%s wire: %s\n", buffer.name.c_str(),
+	        RTLIL::unescape_id(wire->name).c_str());
+#endif
+	    path_delay += buffer.delay;
+	    AddClock(RTLIL::unescape_id(wire->name), wire, clock.Period(),
+	             clock.RisingEdge() + path_delay,
+	             clock.FallingEdge() + path_delay);
+	}
+    }
+}
+
+void Clocks::WriteSdc(std::ostream& file) {
+    for (auto& clock : clocks_) {
+	auto clock_wires = clock.GetClockWires();
+	// FIXME: Input port nets are not found in VPR
+	if (std::all_of(clock_wires.begin(), clock_wires.end(),
+	                [&](RTLIL::Wire* wire) { return wire->port_input; })) {
+	    continue;
+	}
+	file << "create_clock -period " << clock.Period();
+	file << " -waveform {" << clock.RisingEdge() << " "
+	     << clock.FallingEdge() << "}";
+	for (auto clock_wire : clock_wires) {
+	    if (clock_wire->port_input) {
+		continue;
+	    }
+	    file << " " << Clock::ClockWireName(clock_wire);
+	}
+	file << std::endl;
+    }
+}
+
+Clock::Clock(const std::string& name, RTLIL::Wire* wire, float period,
+             float rising_edge, float falling_edge)
+    : name_(name),
+      period_(period),
+      rising_edge_(rising_edge),
+      falling_edge_(falling_edge) {
+    AddWire(wire);
+}
+
+Clock::Clock(const std::string& name, std::vector<RTLIL::Wire*> wires,
+             float period, float rising_edge, float falling_edge)
+    : name_(name),
+      period_(period),
+      rising_edge_(rising_edge),
+      falling_edge_(falling_edge) {
+    std::for_each(wires.begin(), wires.end(),
+                  [&, this](RTLIL::Wire* wire) { AddWire(wire); });
+}
+
+void Clock::AddWire(RTLIL::Wire* wire) {
+    auto clock_wire = std::find(clock_wires_.begin(), clock_wires_.end(), wire);
+    if (clock_wire == clock_wires_.end()) {
+	clock_wires_.push_back(wire);
+    }
+}
+
+void Clock::UpdateClock(RTLIL::Wire* wire, float period, float rising_edge,
+                        float falling_edge) {
+    UpdateWires(wire);
+    UpdatePeriod(period);
+    UpdateWaveform(rising_edge, falling_edge);
+}
+
+void Clock::UpdateWires(RTLIL::Wire* wire) {
+    if (std::find(clock_wires_.begin(), clock_wires_.end(), wire) ==
+        clock_wires_.end()) {
+	clock_wires_.push_back(wire);
+    }
+}
+
+void Clock::UpdatePeriod(float period) {
```

Comment:
This is assuming a 50% duty cycle.  Did you check this assumption when reading in the clock constraint?

Reply:
When I am looking at this now I come to think that we don't really want to change the waveform based on the period assuming 50% duty cycle?

Reply:
I actually think this particular bit of code is right, but maybe by accident.  If when creating a clock, you only specify the period, then the duty cycle is assumed to be 50%, which is what this code does.  Then if you also specify the waveform, it updates the rising edge and falling edge times.

Reply:
Check is done on waveform update.